### DISCUSSION TITLE
Audio playback (AVPlayer, AVPlayerItem, etc) support...

### DISCUSSION THREAD

**Discussion by jeffc-dev** [skiptools/skip#158] [Ideas]
I noticed that there is a `skip-av` framework that seems to have implemented a small subset of `AVKit` for use by the VideoPlayer component.

Are there any plans for a more full-featured AVKit implementation, particularly for use in playing _audio_ content?

My main goal for Skip in the near term is to create an app that can play on-demand audio content (e.g. podcast MP3s) as well as live HLS audio streams, but it doesn't look like Skip currently has native support for anything like that. 

I could come up with a more detailed list, but typically the above functionality would require support in the following areas:

- `AVPlayer` (e.g. `automaticallyWaitsToMinimizeStalling`, `currentItem`, `timeControlStatus`, `replaceCurrentItem`, etc).
- `AVPlayerItem` (e.g. `status`, `cancelPendingSeeks`, `currentTime()`, `currentDate()`).
- `AVAudioSession` (e.g. `setCategory` and `setActive`).
- `AVPlayer`/`AVPlayerItem` observation functionality (e.g. `addPeriodicTimeObserver`, `observe`).
- `MediaPlayer` for Control Center support (e.g. `MPNowPlayingInfoCenter`, `MPRemoteCommandCenter`).

It's possible that a great deal of this has no real equivalent on Android and isn't strictly necessary to support in any _direct_ way, but this should offer a basic idea of what I'm looking to do: Play media (MP3 or HLS) via URL, see where the user is in relative playback terms, handle playback/stall events, update system media controls, etc.

The `skip-av` framework hasn't been committed to in ~7 months. Are there any current plans to expand its support beyond the `VideoPlayer` functionality it currently supports?

FWIW: I'm fine handling the UX side of all this on my own, so I don't need media controls or whatever. I can perform seeks, play/pause, etc. using my own UI elements. I'm just hoping to avoid having to dig into implementing ExoPlayer from scratch if it can be avoided.

I'd also be willing/able to dogfood all of this in a pretty extensive way, if it's something you're interested in supporting. I've worked with AVPlayer quite a bit in the past.

Thanks!

**Comment by marcprux** [comment]
You are correct that [SkipAV](https://github.com/skiptools/skip-av/blob/main/Sources/SkipAV/SkipAV.swift) is a very minimal example of bridging the [AVPlayer](https://developer.apple.com/documentation/avfoundation/media_playback) API to Android's [Exoplayer](https://developer.android.com/media/media3/exoplayer). We've basically been waiting for someone to come along with needs like yours to justify the effort in enhancing the API to support other use cases. 😃 

We ourselves have minimal experience with the media APIs on either platform, so your best bet might be to fork the [skip-av](https://github.com/skiptools/skip-av) repository and explore filling in the Kotlin APIs that you need, as per the [contribution guide](https://skip.tools/docs/contributing/). Some of them might be straightforward, but you'll need to explore the API space to see which types and functions can be logically mapped between the two targets. For example, is [AVPlayerItem](https://developer.apple.com/documentation/avfoundation/avplayeritem) really a 1-1 match with [MediaItem](https://developer.android.com/reference/androidx/media3/common/MediaItem)? If so, then many of the function and property implementations will be straightforward. If not, then you may need to map new types on the Kotlin side that perhaps handle aggregating multiple types together to provide the same API shape as AVKit.

Or, if it turns out that the two APIs are so dissimilar that it doesn't make sense to implement the Exoplayer as an AVPlayer, then you may need to just come up with a new type that encompasses both. It can sometimes be illustrative to see how other cross-platform tools like Flutter implement these sorts of abstractions on the [iOS](https://github.com/flutter/packages/tree/main/packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation) and [Android](https://github.com/flutter/packages/tree/main/packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer) sides.

This might sound like a lot of work, but getting started is fairly easy: just start implementing `#if SKIP` blocks in your skip-av fork or clone, and you should be able to get some quick wins. Please do let us know if we can help with the process. Despite not having much AVKit/Exoplayer expertise ourselves, we'll be more than happy to provide guidance and assistance with some of the tricky parts of fitting together the two APIs.
